### PR TITLE
[RLlib] Cleanup examples folder (new API stack) vol 31: Add hierarchical training example script.

### PR DESCRIPTION
**Pull request by sven1977**
Cleanup examples folder (new API stack) vol 31:
* Add new example env (six room env) with two versions: flat and hierarchical.
* Add hierarchical training example script.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/rllib/rllib-examples.rst [hunk 1]
```diff
@@ -186,6 +186,13 @@ GPU (for Training and Sampling)
    with performance improvements during evaluation.
 
 
+Hierarchical Training
++++++++++++++++++++++
+
+- `Policy Inference after Training <https://github.com/ray-project/ray/blob/master/rllib/examples/inference/policy_inference_after_training.py>`__:
```
Comment (sven1977) [inline doc/source/rllib/rllib-examples.rst:192]: sorry, still WIP ... please ignore this change here for now

File: rllib/env/utils/infinite_lookback_buffer.py [hunk 2]
```diff
@@ -533,9 +533,18 @@ def _get_int_index(
     ):
         data_to_use = self.data
         if _ignore_last_ts:
-            data_to_use = self.data[:-1]
+            if self.finalized:
```
Comment (sven1977) [inline rllib/env/utils/infinite_lookback_buffer.py:536]: these were all bugs

File: rllib/env/multi_agent_episode.py [hunk 3]
```diff
@@ -2532,12 +2532,15 @@ def _get_single_agent_data_by_index(
...
                 # buffers.
                 if extra_model_outputs_key is None:
+                    assert hanging_val is None or isinstance(hanging_val, dict)
                     return {
                         key: sub_buffer.get(
                             indices=index_incl_lookback - sub_buffer.lookback,
                             neg_index_as_lookback=True,
                             fill=fill,
-                            _add_last_ts_value=hanging_val,
+                            _add_last_ts_value=(
```
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:2541]: another bug fix

File: doc/source/rllib/rllib-examples.rst [hunk 4]
```diff
@@ -186,6 +186,16 @@ GPU (for Training and Sampling)
    with performance improvements during evaluation.
 
 
+Hierarchical Training
++++++++++++++++++++++
+
+- `Hierarchical RL Training <https://github.com/ray-project/ray/blob/master/rllib/examples/hierarchical/hierarchical_training.py>`__:
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-examples.rst:192]: Awesome!!

File: doc/source/rllib/rllib-examples.rst [hunk 5]
```diff
@@ -186,6 +186,16 @@ GPU (for Training and Sampling)
...
+   Showcases a hierarchical RL setup inspired by automatic subgoal discovery and subpolicy specialization. A high-level policy selects subgoals and assigns one of three
+   specialized low-level policies to achieve them within a time limit, encouraging specialization and efficient task-solving.
+   The agent has to navigate a complex grid-world environment. The example highlights the advantages of hierarchical
+   learning over flat approaches by demonstrating significantly improved learning performance in challenging, goal-oriented tasks.
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-examples.rst:196]: This is just really interesting

File: rllib/env/multi_agent_episode.py [hunk 6]
```diff
@@ -551,8 +551,8 @@ def add_env_step(
                     # duplicate the previous one (this is a technical "fix" to properly
                     # complete the single agent episode; this last observation is never
                     # used for learning anyway).
-                    _observation = sa_episode.get_observations(-1)
-                    _infos = sa_episode.get_infos(-1)
+                    _observation = sa_episode._last_added_observation
```
Comment (simonsays1980) [inline rllib/env/multi_agent_episode.py:554]: Sweet!

File: rllib/examples/hierarchical/hierarchical_training.py [hunk 7]
```diff
@@ -1,152 +1,170 @@
...
+We use the `SixRoomEnv` and `HierarchicalSixRoomEnv`, both sharing the same built-in
+maps. The envs are similar to the FrozenLake-v1 env, but support walls (inner and outer)
+through which the agent cannot walk.
+
+
+How to run this script
+----------------------
+`python [script file name].py --enable-new-api-stack --map=large --time-limit=50`
+
+Use the `--flat` option to disable the hierarchical setup and learn the simple (flat)
```
Comment (simonsays1980) [inline rllib/examples/hierarchical/hierarchical_training.py:41]: This is so cool!

**Review by simonsays1980 — APPROVED** [review]
LGTM. This is one of the coolest examples we will have in our stack!

**Pull request merged**